File: 91.R
# Right triangles with integer coordinates

Args <- commandArgs()
max  <- ifelse(is.na(Args[6]), 50, as.integer(Args[6]))

ints <- seq.int(0, max)
df <- expand.grid(x1=ints, y1=ints, x2=ints, y2=ints)
df <- df[!(df$x1 == 0 & df$y1 == 0),]
df <- df[!(df$x2 == 0 & df$y2 == 0),]
df <- df[!(df$x1 == df$x2 & df$y1 == df$y2),]
df$x3 <- df$x2 - df$x1
df$y3 <- df$y2 - df$y1
df$p1 <- df$x1 * df$x2 + df$y1 * df$y2
df$p2 <- df$x2 * df$x3 + df$y2 * df$y3
df$p3 <- df$x3 * df$x1 + df$y3 * df$y1
df <- df[!df$p1 | !df$p2 | !df$p3,]

cat(nrow(df) / 2, fill=TRUE)

File: 91.js
#!/usr/bin/env node

// Right triangles with integer coordinates

var program = require('commander');

program
    .version('0.1.0')
    .description('Right triangles with integer coordinates')
    .option('-m, --max <int>', 'The maximum x or y coordinate', Number, 50)
    .parse(process.argv);

var count = 0;

for (var x1 = 0; x1 <= program.max; x1++) {
    for (var y1 = 0; y1 <= program.max; y1++) {
        if (x1 == 0 && y1 == 0) {
            continue;
        }
        var op = [x1, y1];
        for (var x2 = 0; x2 <= program.max; x2++) {
            for (var y2 = 0; y2 <= program.max; y2++) {
                if (x2 == 0 && y2 == 0) {
                    continue;
                }
                if (x1 == x2 && y1 == y2) {
                    continue;
                }
                var oq = [x2, y2];
                var pq = [x2 - x1, y2 - y1];
                if (!dot_product(op, oq) ||
                    !dot_product(op, pq) ||
                    !dot_product(oq, pq)) {
                    count++;
                }
            }
        }
    }
}

console.log(count / 2);

function dot_product(v1, v2) {
    return v1[0] * v2[0] + v1[1] * v2[1];
}
